### PR TITLE
Complete implementation of PNG IHDRC chunk decoding

### PR DESCRIPTION
**Pull request by piponazo**
I started digging into the PNG specification and I will create few PRs dedicated to improve the decoding of PNG chunks. 

In this first PR I just focus on the [IHDR chunk](http://www.libpng.org/pub/png/spec/1.2/png-1.2-pdg.html#C.IHDR). We were only reading the width and height values from there and ignoring the rest of values. Now we read all the fields and we do the pertinent checks on their values. This should make the code more robust against malicious handmade PNG files. 

I also took the opportunity to continue updating the doxygen style and clang-format the files touched. 

Recommendation: review commit by commit.

### CODE REVIEW

**Comment by D4N** [comment]
@piponazo Could you rebase this on master so that we get a coverage report?

**Comment by piponazo** [comment]
> @piponazo Could you rebase this on master so that we get a coverage report?

done

**Review by clanmills — COMMENTED** [review]
This looks like a good overhaul of the PNG code.  And that's a good thing.  Thanks for digging around in that.

I am concerned about the symbol EXV_HAVE_PNG as that implies we are using libpng.  Perhaps EXV_ENABLE_PNG is a better name which (based on EXIV2_ENABLE_PNG in CMakeLists.txt).

For that matter, perhaps it's worthwhile replacing EXV_this_that with EXIV2_this_that.

**Comment by piponazo** [comment]
> I am concerned about the symbol EXV_HAVE_PNG as that implies we are using libpng. Perhaps EXV_ENABLE_PNG is a better name which (based on EXIV2_ENABLE_PNG in CMakeLists.txt). For that matter, perhaps it's worthwhile replacing EXV_this_that with EXIV2_this_that.

I totally agree with you. I'll make both changes 😉

File: src/pngchunk_int.hpp [hunk 1]
```diff
@@ -20,186 +20,136 @@
...
+            /// @brief Return a compressed (zTXt) or uncompressed (tEXt) PNG ASCII text chunk (length + chunk type +
+            /// chunk data + CRC) as a string.
+            /// @param keyword  Keyword for the PNG text chunk
+            /// @param text     Text to be recorded in the PNG chunk.
+            /// @param compress Flag indicating whether to compress the PNG chunk data.
+            /// @return String containing the PNG chunk
+            static std::string makeAsciiTxtChunk(const std::string& keyword, const std::string& text, bool compress);
+
+            /// @brief Return a compressed or uncompressed (iTXt) PNG international text chunk (length + chunk type +
+            /// chunk data + CRC) as a string.
```
Comment (D4N) [inline src/pngchunk_int.hpp]:
```suggestion
            ///     chunk data + CRC) as a string.
```

**Review by D4N — CHANGES_REQUESTED** [review]
I've added some nagging to the documentation, but unfortunately realized too late that you only reformated most of the doc comments, so feel free to ignore the "please clarify this" comments, unless you're highly motivated to address them :wink:.

I'd suggest to throw `kerFailedToReadMetadata` from the function that you completed, for consistency with the other png reading functions (that would also make the changes in the system test obsolete).

File: include/exiv2/pngimage.hpp [hunk 2]
```diff
@@ -40,48 +40,32 @@
...
-              method to get a temporary reference.
-          @param create Specifies if an existing image should be read (false)
-              or if a new file should be created (true).
-         */
+        /// @brief Constructor that can either open an existing PNG image or create a new image from scratch.
+        ///
+        /// If a new image is to be created, any existing data is overwritten. Since the constructor can not return a
+        /// result, callers should check the good() method after object construction to determine success or failure.
+        /// @param io An auto-pointer that owns a BasicIo instance used for reading and writing image metadata.
+        /// \b Important: The constructor takes ownership of the passed in BasicIo instance through the auto-pointer.
```
[the remark this answers is not in the document]
Comment (piponazo) [reply include/exiv2/pngimage.hpp]: I agree, I saw some of these comments and I actually simplified them (not sure if it was done in this PR or in other).

File: include/exiv2/pngimage.hpp [hunk 3]
```diff
@@ -40,48 +40,32 @@
...
         PngImage(BasicIo::UniquePtr io, bool create);
         //@}
 
         //! @name Manipulators
         //@{
         void readMetadata() override;
         void writeMetadata() override;
-
-        /*!
-          @brief Print out the structure of image file.
```
[the remark this answers is not in the document]
Comment (piponazo) [reply include/exiv2/pngimage.hpp]: It is exactly the same comment that comes from the Base class (**Image**). In Doxygen, there is documentation inheritance :smile_cat:

File: src/pngchunk_int.hpp [hunk 4]
```diff
@@ -20,186 +20,136 @@
...
+            /// @param metadata    metadata buffer.
+            /// @param type        metadata type.
+            static std::string makeMetadataChunk(const std::string& metadata, MetadataId type);
+
+        private:
+            /// @brief Parse PNG Text chunk to determine type and extract content.
+            /// Supported Chunk types are tTXt, zTXt, and iTXt.
+            static DataBuf parseTXTChunk(const DataBuf& data, int keysize, TxtChunkType type);
+
+            /// @brief Parse PNG chunk contents to extract metadata container and assign it to image.
```
[the remark this answers is not in the document]
Comment (piponazo) [reply src/pngchunk_int.hpp]: I'll try to do it when I write unit tests for it in the future :wink:

**Comment by piponazo** [comment]
> I'd suggest to throw kerFailedToReadMetadata from the function that you completed, for consistency with the other png reading functions (that would also make the changes in the system test obsolete).

I am happy to apply this change, although the truth is that there is an existing mix in that file of exceptions: `kerFailedToReadImageData` and `kerCorruptedMetadata`. Do you know the difference between those 2 values ? When we should use one or the other ? In case there is not so much difference I would vote for removing one of them (In another PR).

**Comment by piponazo** [comment]
> For that matter, perhaps it's worthwhile replacing EXV_this_that with EXIV2_this_that.

@clanmills I just replaced the definition I changed in this file. Since to do that change for all the definitions is a quite big task, I would prefer to do it in another PR.

**Comment by clanmills** [comment]
Changing EXV_HAVE_PNG to EXIV2_ENABLE_PNG is the change I requested.

Changing all those EXV_ defines is something for another day.  We could ask Nehal to do that when everybody's on vacation (July?) and the change slides into the code without friction.

**Review by D4N — APPROVED** [review]

**Comment by D4N** [comment]
> > I'd suggest to throw kerFailedToReadMetadata from the function that you completed, for consistency with the other png reading functions (that would also make the changes in the system test obsolete).
> 
> I am happy to apply this change, although the truth is that there is an existing mix in that file of exceptions: `kerFailedToReadImageData` and `kerCorruptedMetadata`. Do you know the difference between those 2 values ? When we should use one or the other ? In case there is not so much difference I would vote for removing one of them (In another PR).

I think `kerFailedToReadImageData` existed a long time before `kerCorruptedMetadata`, which I might have added when I just started contributing ("back then" we just had error codes and no enum for the numbers, so new errors got added left right and center, see #323)

**Pull request merged**